src/asynch/wifi.rs

```diff
@@ -0,0 +1,308 @@
+        inner.process_urc_messages();
+        JoinState {
+            connected: inner.session.joined,
+            ip_assigned: inner.session.ip_assigned,
+        }
+    }
+
+    async fn get_address(&mut self) -> Result<LocalAddress, AddressErrors> {
+        let responses = self.inner.lock().await.send_command(ObtainLocalAddressCommand::new()).await?;
+        LocalAddress::from_responses(responses)
+    }
+
+    async fn set_auto_connect(&mut self, enabled: bool) -> Result<(), CommandError> {
+        self.inner.lock().await.send_command(AutoConnectCommand::new(enabled)).await?;
+        Ok(())
+    }
+
+    async fn restart(&mut self) -> Result<(), CommandError> {
+        let mut inner = self.inner.lock().await;
+        inner.session.ready = false;
+        inner.send_command(RestartCommand::default()).await?;
+
+        inner.session = Session::default();
+
+        let task = async {
+            while !inner.session.ready {
+                inner.process_urc_messages();
+                yield_now().await;
+            }
+        };
+
+        let res = embassy_futures::select::select(
+            Timer::after_secs(5),
+            task,
+        ).await;
+
+        match res {
+            Either::First(_) => Err(CommandError::ReadyTimeout),
+            Either::Second(_) => Ok(())
+        }
+    }
+}
+
+impl<
+        'urc_sub,
+        A: AtatClient,
+        const TX_SIZE: usize,
+        const RX_SIZE: usize,
+        const URC_CAPACITY: usize,
+    > Adapter<'urc_sub, A, TX_SIZE, RX_SIZE, URC_CAPACITY>
+{
+    pub fn new(
+        client: A,
+        urc_subscription: UrcSubscription<'urc_sub, URCMessages<RX_SIZE>, URC_CAPACITY, 1>,
+        send_timeout: Duration,
+    ) -> Self {
+        Self {
+            inner: Mutex::new(InnerAdapter::new(client, urc_subscription, send_timeout)),
+        }
+    }
+}
+
+impl<
+    'urc_sub,
+    A: AtatClient,
+    const TX_SIZE: usize,
+    const RX_SIZE: usize,
+    const URC_CAPACITY: usize,
+> TcpConnect for Adapter<'urc_sub, A, TX_SIZE, RX_SIZE, URC_CAPACITY>
+{
+    type Error = Error;
+
+    type Connection<'a> = Connection<'a, 'urc_sub, A, TX_SIZE, RX_SIZE, URC_CAPACITY>
+        where
+            Self: 'a;
+
+    async fn connect<'a>(&'a self, remote: SocketAddr) -> Result<Self::Connection<'a>, Self::Error> {
+        let mut inner = self.inner.lock().await;
+        inner.enable_multiple_connections().await?;
+        let socket = inner.open_socket()?;
+        inner.process_urc_messages();
+
+        if inner.session.is_socket_connected(&socket) {
+            return Err(Error::AlreadyConnected);
+        }
+
+        inner.enable_passive_receiving_mode().await?;
+        inner.session.already_connected = false;
+
+        let command = match remote {
+            SocketAddr::V4(address) => ConnectCommand::tcp_v4(socket.link_id, address),
+            SocketAddr::V6(address) => ConnectCommand::tcp_v6(socket.link_id, address),
+        };

```

Comment:
What about implementing The `From` Trait on `ConnectCommand` an calling it then both in sync and async version?